continue fact gathering even without dmidecode

toshywoshy opened the pull request:
##### SUMMARY
Bugfix Pull Request

If dmidecode is not available we still wan to continue with fact gathering.
On certain platforms dmidecode won't even be available

##### COMPONENT NAME
setup ( fact gathering )

##### ANSIBLE VERSION

```
ansible 2.5.0 (dmidecode-na 06a1ed6718) last updated 2018/01/16 12:49:32 (GMT +200)
  config file = /home/ansible/.ansible.cfg
  configured module search path = [u'/home/ansible/ansible/library']
  ansible python module location = /home/ansible/ansible/lib/ansible
  executable location = /home/ansible/ansible/bin/ansible
  python version = 2.7.5 (default, May 16 2016, 12:54:51) [GCC 4.8.5 20150623 (PowerEL 4.8.5-10)]
```

toshywoshy committed:
continue fact gathering even without dmidecode

If dmidecode is not available we still wan to continue with fact
gathering.
On certain platforms dmidecode just won't work

Signed-off-by: Toshaan Bharvani <tab@toshaan.com>

Akasurde reviewed (commented):
LGTM

alikins commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -223,13 +223,15 @@ def get_virtual_facts(self):
-            if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
-                virtual_facts['virtualization_type'] = 'VMware'
-                virtual_facts['virtualization_role'] = 'guest'
+        # We still want to continue even if dmidecode is not available
```
Patch looks good as is. So :shipit: for me.

But was thinking it might be a good time to move this to a helper method, since then it could return early and lose a few indention levels if you wanted to.

``` python

def _collect_dmi(self):
    dmi_virt_facts = {}

    dmi_bin = self.module.get_bin_path('dmidecode')
    
    if not dmi_bin:
        return dmi_virt_facts
    
    (rc, out, err) = self.module.run_command('%s -s system-product-name' % dmi_bin)
    
    if rc != 0:
        return dmi_virt_facts

    # I might split this into two methods, one for finding and running dmi, and another for parsing it.
    # and passing in the dmidecode output to the parsing method
    # that would make the parsing code easy to test
    vendor_name = ''.join([line.strip() for line in out.splitlines() if not line.startswith('#')])
    
    if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
        dmi_virt_facts['virtualization_type'] = 'VMware'
        dmi_virt_facts['virtualization_role'] = 'guest'

    return dmi_virt_facts

```

with
``` python

dmi_virt_facts = self._collect_dmi()
virtual_facts.update(dmi_virt_facts)

```

or if split into two

``` python

dmi_bin_output = self._run_dmidecode()
dmi_virt_facts = self._collect_dmi(dmi_bin_output)
virtual_facts.update(dmi_virt_facts)

Akasurde replied:
I have one more suggestion, it would be good to replace

```
if vendor_name in ['VMware Virtual Platform', 'VMware7,1']:
```

with 

```
if vendor_name.startwith('VMware'):
```

I am not sure about `VMware 7,1` uniqueness on other ESXi version. (I missed this in last commit)

toshywoshy replied:
@alikins IMHO that would only make sense if all the collections would follow the same logic in separate functions 
@Akasurde I agree and have updated the patch

MartinNowak replied:
This introduced a regression, it's called start**s**with in Python.
Also this doesn't (and didn't) return early like all the other branches, so the VMware result would always be overwritten by NAs (see https://github.com/ansible/ansible/pull/34132#pullrequestreview-91756340).

alikins reviewed (approved):
Looks good to me. Have a suggestion that might make it easier to test in the future if interested.

toshywoshy committed:
changed VMware lookup

As suggested by Akasurde, have also changed the lookup for VMware
The current check only works with a single version, now it should be
more generic

Signed-off-by: Toshaan Bharvani <tab@toshaan.com>

Akasurde commented:
@toshywoshy Are you still working on this ?

toshywoshy commented:
@Akasurde I rebased already as the bot told me, now someone needs to approve and merge this, as all test pass and everything else is good to go
I see that OpenStack  also want this merged in, but somehow that didn't make it go any faster
I already pinged @dagwieers to look at it, but you are welcome to approve, unless you have some comments on the changes

kevlar1818 commented:
Any status update on getting this merged? This seems like a show-stopper for a fair amount of people (myself included).

alikins commented:
lets merge it and add tests later I guess.

Akasurde commented:
@alikins OK.

core-process commented:
My temporary workaround (Ubuntu Artful, Ansible PPA):

```sh
cd
wget https://github.com/ansible/ansible/commit/0ad783c381fc811ca0b375ceba83eeab1914a08c.patch

cd /usr/lib/python2.7/dist-packages/
sudo patch -p2 -i ~/0ad783c381fc811ca0b375ceba83eeab1914a08c.patch

cd ansible/module_utils/facts/virtual
sudo python -m compileall linux.py

cd
rm 0ad783c381fc811ca0b375ceba83eeab1914a08c.patch
```
